sa/sysvars.go

```diff
@@ -0,0 +1,219 @@
+package sa
+
+import (
+	"fmt"
+	"regexp"
+)
+
+var (
+	checkStringQuoteRE = regexp.MustCompile(`^'[0-9A-Za-z_\-=:]+'$`)
+	checkIntRE         = regexp.MustCompile(`^\d+$`)
+	checkImproperIntRE = regexp.MustCompile(`^'\d+'$`)
+	checkNumericRE     = regexp.MustCompile(`^\d+(\.\d+)?$`)
+	checkBooleanRE     = regexp.MustCompile(`^([0-1])|(?i)(true|false)|(?i)(on|off)`)
+)
+
+// checkMariaDBSystemVariables validates a MariaDB config passed in via SA
+// setDefault or DSN. This manually curated list of system variables was
+// partially generated by a tool in issue #6687. An overview of the validations
+// performed are:
+//
+//   - Correct quoting for strings and string enums prevent future
+//     problems such as PR #6683 from occurring.
+//
+//   - Regex validation is performed for the various booleans, floats, integers, and strings.
+//
+// Only session scoped variables should be included. A session variable is one
+// that affects the current session only. Passing a session variable that only
+// solely in the global scope causes database connection error 1045.
+// https://mariadb.com/kb/en/set/#global-session
+func checkMariaDBSystemVariables(name string, value string) error {
+	// System variable names will be indexed into the appropriate hash sets
+	// below and can possibly exist in several sets.
+
+	// Check the list of currently known MariaDB string type system variables
+	// and determine if the value is a properly formatted string e.g.
+	// sql_mode='STRICT_TABLES'
+	mariaDBStringTypes := map[string]struct{}{
+		"character_set_client":           {},
+		"character_set_connection":       {},
+		"character_set_database":         {},
+		"character_set_filesystem":       {},
+		"character_set_results":          {},
+		"character_set_server":           {},
+		"collation_connection":           {},
+		"collation_database":             {},
+		"collation_server":               {},
+		"debug_sync":                     {},
+		"enforce_storage_engine":         {},
+		"external_user":                  {},
+		"lc_time_names":                  {},
+		"old_mode":                       {},
+		"optimizer_switch":               {},
+		"proxy_user":                     {},
+		"session_track_system_variables": {},
+		"sql_mode":                       {},
+		"time_zone":                      {},
+	}
+
+	if _, found := mariaDBStringTypes[name]; found {
+		if checkStringQuoteRE.FindString(value) != value {
+			return fmt.Errorf("%v=%v string is not properly quoted", name, value)
```

Comment:
```suggestion
			return fmt.Errorf("%s=%s string is not properly quoted", name, value)
```

In general I expect %s when the parameter is known to be a string. Otherwise, someone reading the code might wonder "wait, what other types can `name` and `value` have?"

(also instances of this below)